proposals/break-continue-in-inline-lambdas.md

```diff
@@ -0,0 +1,175 @@
+syntactically.
+
+Also, right now we have such an invariant in Kotlin: a pair of **keywords** `break`/`continue` tied to the closest `while` or
+`for` **keywords**. `return` **keyword** is tied to the closest `fun` **keyword**. The proposed alternative solution breaks this
+invariant.
+
+* **Q**: Why not make labels mandatory for non-local `break`/`continue`? That would eliminate `forEach` puzzler.
+* **A**: Because Kotlin doesn't require labels for non-local `return`. For sake of consistency, We should either require labels
+  for both `break`/`continue` and `return` cases or don't require the labels again for the both cases.
+
+* **Q**: Ok, why not make labels mandatory for both non-local `return` and non-local `break`/`continue` then?
+* **A**: We can't make non-local `return` to be obligatory labeled because it would break inline functions first-class
+  citizenship in language constructs. E.g. we value "return if not null" idiomatic Kotlin construct `nullable?.let { return it }`.
+  If non-local `return` were obligatory labeled then the code would be much more verbose
+  `nullable?.let { return@functionName it }`.
+
+It worth noticing that we don't abandon the idea to provide some "break"/"continue" mechanism for "loop-like" functions
+[KT-19748](https://youtrack.jetbrains.com/issue/KT-19748). It just that we agree that we will never support unlabeled non-local
+`break`/`continue` for "loop-like" functions.
+
+## Tooling for puzzling code
+
+Some users may find constructions containing `break` or `continue` within lambdas passed to "loop-like" inline functions
+such as `forEach` or `takeWhile` puzzling:
+```kotlin
+fun printUntilZero(producer: () -> List<Int>) {
+    while(true) {
+        val list = producer()
+        list.forEach {
+            if (it == 0) break // Does it exit from while or from forEach?
+            println(it)
+        }
+    }
+}
+```
+
+To reduce the risk of confusion, it was proposed to introduce an IDE inspection and a quickfix recommending to use labeled `break`
+and `continue` in lambdas passed to parameters without an `EXACTLY_ONCE` or `AT_MOST_ONCE` contract (potentially "loop-like").
+After the inspection is applied:
+```kotlin
+fun printUntilZero(producer: () -> List<Int>) {
+    myLoop@ while(true) {
+        val list = producer()
+        list.forEach {
+            if (it == 0) break@myLoop
+            println(it)
+        }
+    }
+}
+```
+
+The inspection should not triggerred for `let`, `run`, etc as they have the corresponding contract. There seems to be no need for
+a similar inspection for non-local returns. We assume that they are less confusing.
+
+## Implementation note
+
+Similarly to non-local returns, non-local break/continue will only work for lambdas passed as parameters not marked
+with `noinline` and `crossinline`.
+
+Non-local `break` and `continue` could be emulated using non-local return statements.
+For example, the following code snippets should be equivalent with respect to DFA:
```

Comment:
@strangepleasures sorry, I don't understand what you mean. I don't understand why this section relies on the semantic of `run`. `run` is just used because it's the most canonical way to invoke a lambda.

My point is different: I want that we really don't forget to check that the implementation is identical. And once we check it, we should mention it in the KEEP document. So the phrase "the following code snippets should be equivalent with respect to DFA" will turn into "in the current implementation the following code snippets is equivalent with respect to DFA"

Reply:
No need to mention "the current implementation", because people will be reading this document years later. Also, not every reader knows what DFA is, so the correct phrase would be "Note that the following code snippets are equivalent in terms of their data flow".

Reply:
@nikitabobko but WHY do we need to check that in the first place? To me, this equivalence is just a fun fact, not a requirement. We could have also provided a semantically equivalent program written in different language, but what for?

Reply:
> but WHY do we need to check that in the first place?

We need to check it to try to find potential bugs we have in the current implementation. It's just one more check. What is the point of "it should work like that" phrase in a KEEP document if, in reality, it doesn't work "like that"? Let's please really make sure that it "works like that"

"it works like that" phrase (without "should"), gives one more invariant about the language.